### PR TITLE
OpenAI Function Calling

### PR DESCRIPTION
**Pull request by philippzagar**
# OpenAI Function Calling

## :recycle: Current situation & Problem
As of now, the `SpeziLLMOpenAI` target doesn't provide an option to use the [OpenAI Function Calling mechanism](https://platform.openai.com/docs/guides/function-calling).
This functionality is crucial to advanced LLM use cases, such as `HealthGPT` or `LLMonFHIR`.

Related to https://github.com/StanfordSpezi/SpeziLLM/issues/10

## :gear: Release Notes 
- `SpeziLLMOpenAI` now supports the OpenAI function calling mechanism via a Domain Specific Language, enabling straightforward usage of function calling within Spezi-based applications.

## :books: Documentation
Inline DocC as well as DocC articles written, other documentation adjusted where necessary.

## :white_check_mark: Testing
Unit tests to properly test injection logic of parameters into `LLMFunction`s as well as validation of schema synthization.
No proper UI tests as we would test the OpenAI API and therefore need proper API tokens and OpenAI balance. However, extensive manual testing via the UI test application.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by philippzagar** [comment]
Related to https://github.com/StanfordSpezi/SpeziLLM/issues/10

**Comment by philippzagar** [comment]
@PSchmiedmayer I added an extra build job that compiles the UI test app in Release mode in order to avoid regressions like the Swift / C++ String Interop Issue we had yesterday. Please adjust the merge checks in the repo for the compile job in Debug and Release mode.

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for the great implementation of function calling in this PR @philippzagar! I read the code and documentation and it is great to see how the property-wrapper-based approach turned out and how this is simplifying a lot of the elements that we manually implemented in LLMonFHIR and other applications!

Also great job with the level of detail of the documentation and the extensive test coverage of the injection and parameter logic!

Feel free to merge the PR once the conversations are all resolved 🚀

File: Sources/SpeziLLMOpenAI/FunctionCalling/Helpers/LLMFunctionParameterIntermediateRepresentation.swift
[the remark this answers is not in the document]
Comment (philippzagar) [reply Sources/SpeziLLMOpenAI/FunctionCalling/Helpers/LLMFunctionParameterIntermediateRepresentation.swift:1]: @PSchmiedmayer I should probably put a copyright on it in regards to you and the Apodini project! :D

File: Sources/SpeziLLMOpenAI/Helpers/LLMStreamResult.swift
[the remark this answers is not in the document]
Comment (philippzagar) [reply Sources/SpeziLLMOpenAI/Helpers/LLMStreamResult.swift:1]: Yep I adjusted it a bit in order to be `struct`-based, as the initial LLMonFHIR implementation was a `class` (which from my perspective was just an "quicker" solution at the time)

File: Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunctionParameter.swift [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+
+
+/// Represents an LLM function calling parameter.
+///
+/// The ``LLMFunctionParameter``enables developers to manually specify the conformance of Swift types to the [OpenAI Function calling schema](https://platform.openai.com/docs/guides/function-calling). <!-- markdown-link-check-disable-line -->
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunctionParameter.swift:14]: As discussed in-person, this is resolved, was likely a copy-paste error during the review (or the OpenAI docs were broken?)

File: Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunction.swift [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+//
+// SPDX-FileCopyrightText: 2022 Stanford University and the project authors (see CONTRIBUTORS.md)
+//
+// SPDX-License-Identifier: MIT
+//
+
+
+/// Represents an OpenAI Function that can be called by a GPT-based LLM.
+///
+/// The `LLMFunction` is the Spezi-based implementation of an OpenAI LLM function (or tool): https://platform.openai.com/docs/guides/function-calling <!-- markdown-link-check-disable-line -->
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunction.swift:12]: As discussed in-person, this is resolved, was likely a copy-paste error during the review (or the OpenAI docs were broken?)

File: .github/workflows/build-and-test.yml
[the remark this answers is not in the document]
Comment (philippzagar) [reply .github/workflows/build-and-test.yml:1]: @PSchmiedmayer Good idea! 🚀  I created the following two PRs in the Spezi and BDHG orgs that tackle this issue: https://github.com/StanfordSpezi/.github/pull/45

File: Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunction.swift [hunk 3]
```diff
@@ -73,5 +73,5 @@ public protocol LLMFunction {
     /// The output is automatically injected into the conversational history.
     ///
     /// - Returns: `String`-based output of the function call that is then provided to the LLM.
-    func execute() async throws -> String
+    func execute() async throws -> String?
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunction.swift:81]: @PSchmiedmayer Added proper docs, good catch 👍

File: Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunction.swift [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+///
+/// Upon initializing the ``LLMOpenAI``, developers can pass an array of ``LLMFunction``s via ``LLMOpenAI/init(parameters:modelParameters:_:)``.
+/// These functions are then made available to OpenAI's GPT models and can be called if the model decides to do so, based on the current conversational context.
+/// An ``LLMFunction`` can have multiple ``LLMFunction/Parameter``s (`@Parameter`) to tailor the requested functionality of the LLM.
+///
+/// The crucial properties of a ``LLMFunction`` are the ``LLMFunction/name``, serving as the identifier of the function, as well as the ``LLMFunction/description``,
+/// enabling the LLM to understand the purpose of the available function call.
+/// The actual logic that is executed upon a function call by the LLM resides in the ``LLMFunction/execute()`` function.
+/// ``LLMFunction/execute()`` returns a `String?`, containing all the information the Spezi application wants to provide to the LLM upon a function call,
+/// and is automatically injected into the LLM conversation by `SpeziLLM`. In case of a `nil` return value, the function call completion is automatically injected into the LLM conversation.
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunction.swift]: Added proper docs as indicated above

**Pull request merged**